File: openmrs-client/src/main/java/org/openmrs/mobile/models/typeConverters/FormResourceConverter.java
/*
 * The contents of this file are subject to the OpenMRS Public License
 * Version 1.0 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://license.openmrs.org
 *
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
 * License for the specific language governing rights and limitations
 * under the License.
 *
 * Copyright (C) OpenMRS, LLC.  All Rights Reserved.
 */

package org.openmrs.mobile.models.typeConverters;

import androidx.room.TypeConverter;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;

import org.openmrs.mobile.databases.entities.FormResourceEntity;

import java.io.Serializable;
import java.lang.reflect.Modifier;
import java.lang.reflect.Type;
import java.util.List;

public class FormResourceConverter implements Serializable {

    @TypeConverter
    public static List<FormResourceEntity> fromString(String value) {
        Type listType = new TypeToken<List<FormResourceEntity>>() {}.getType();
        GsonBuilder builder = new GsonBuilder();
        builder.excludeFieldsWithModifiers(Modifier.FINAL, Modifier.TRANSIENT, Modifier.STATIC);
        Gson gson = builder.create();
        List<FormResourceEntity> formResources = gson.fromJson(value, listType);
        return formResources;
    }

    @TypeConverter
    public static String listToString(List<FormResourceEntity> list) {
        GsonBuilder builder = new GsonBuilder();
        builder.excludeFieldsWithModifiers(Modifier.FINAL, Modifier.TRANSIENT, Modifier.STATIC);
        Gson gson = builder.create();
        return gson.toJson(list);
    }
}

File: openmrs-android-sdk/src/main/java/com/openmrs/android_sdk/library/models/typeConverters/FormResourceConverter.kt
/*
 * The contents of this file are subject to the OpenMRS Public License
 * Version 1.0 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://license.openmrs.org
 *
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
 * License for the specific language governing rights and limitations
 * under the License.
 *
 * Copyright (C) OpenMRS, LLC.  All Rights Reserved.
 */
package com.openmrs.android_sdk.library.models.typeConverters

import androidx.room.TypeConverter
import com.openmrs.android_sdk.library.databases.entities.FormResourceEntity
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import java.io.Serializable
import java.lang.reflect.Modifier

object FormResourceConverter : Serializable {
    @JvmStatic
    @TypeConverter
    fun fromString(value: String?): List<FormResourceEntity> {
        val listType = object : TypeToken<List<FormResourceEntity?>?>() {}.type
        val builder = GsonBuilder()
        builder.excludeFieldsWithModifiers(Modifier.FINAL, Modifier.TRANSIENT, Modifier.STATIC)
        val gson = builder.create()
        return gson.fromJson(value, listType)
    }

    @JvmStatic
    @TypeConverter
    fun listToString(list: List<FormResourceEntity?>?): String {
        val builder = GsonBuilder()
        builder.excludeFieldsWithModifiers(Modifier.FINAL, Modifier.TRANSIENT, Modifier.STATIC)
        val gson = builder.create()
        return gson.toJson(list)
    }
}